<b>Шаг 1. Обработка данных.</b> Подгружаем библиотеку <b>pandas</b> для анализа данных. Устанавливаем макс. кол-во выводимых строк равным 10, снимаем ограничение на вывод всех столбцов.

In [193]:
import pandas as pd
pd.set_option('display.max_rows', 10)
pd.set_option('display.max_columns', None)

На основе .csv файлов создаем dataframe для работы данными.

In [194]:
df1 = pd.read_csv('raw_data1.csv')
df2 = pd.read_csv('raw_data2.csv')
df3 = pd.read_csv('raw_data3.csv')

Конкатенируем строки

In [195]:
df = pd.concat([df1,df2,df3])

In [196]:
df.shape

(14847, 34)

In [197]:
df

Unnamed: 0       отделка  год постройки жилая площадь корпус, строение  \
0              0           NaN            NaN           NaN              NaN   
1              1           NaN            NaN           NaN              NaN   
2              2   без отделки            NaN           NaN          дом 4.1   
3              3   без отделки            NaN           NaN          дом 2.2   
4              4           NaN            NaN           NaN              NaN   
...          ...           ...            ...           ...              ...   
4919        4919           NaN            NaN           NaN              NaN   
4920        4920           NaN         2021.0       11.3 м²              NaN   
4921        4921           NaN         1970.0       43.1 м²              NaN   
4922        4922           NaN            NaN         12 м²              NaN   
4923        4923           NaN            NaN         32 м²              NaN   

                ремонт      тип комнат общая площадь высота потолков  \
0      требует ремонта             NaN       37.4 м²             NaN   
1        косметический   изолированные       57.9 м²             NaN   
2                  NaN             NaN       36.8 м²             NaN   
3                  NaN   изолированные         74 м²             NaN   
4        косметический   изолированные         48 м²             NaN   
...                ...             ...           ...             ...   
4919              евро   изолированные       53.4 м²             NaN   
4920              евро             NaN       39.1 м²           2.6 м   
4921     косметический   изолированные       58.7 м²             NaN   
4922     косметический             NaN       33.4 м²             NaN   
4923      дизайнерский   изолированные         66 м²           2.6 м   

     пассажирский лифт количество комнат балкон или лоджия площадь кухни  \
0                  NaN                 1            балкон       10.7 м²   
1                  NaN                 3            балкон          5 м²   
2                  NaN                 1               NaN           NaN   
3                  NaN                 2               NaN       19.3 м²   
4                  NaN                 2            балкон          9 м²   
...                ...               ...               ...           ...   
4919               NaN                 2               NaN        8.3 м²   
4920                 2                 1               NaN       19.7 м²   
4921               NaN                 4            балкон        5.9 м²   
4922               NaN            студия            балкон           NaN   
4923               NaN                 2            лоджия         30 м²   

     в доме грузовой лифт                             официальный застройщик  \
0       NaN           NaN                                                NaN   
1       NaN           NaN                                                NaN   
2       NaN           NaN                             ооо «сз «южный парк-4»   
3       NaN           NaN   ооо «специализированный застройщик «времена г...   
4       NaN           NaN                                                NaN   
...     ...           ...                                                ...   
4919    NaN           NaN                                                NaN   
4920    NaN           нет                                                NaN   
4921    NaN           NaN                                                NaN   
4922    NaN           NaN                                                NaN   
4923    NaN           NaN                                                NaN   

            срок сдачи         вид сделки тёплый пол техника мебель  \
0                  NaN   возможна ипотека        NaN     NaN    NaN   
1                  NaN   возможна ипотека        NaN     NaN    NaN   
2      2 кв. 2025 года                NaN        NaN     NaN    NaN   
3      1 кв. 2024 года

Удаляем колонку <b>Unnamed:0</b>

In [198]:
df.drop(['Unnamed: 0'], axis=1,inplace=True)

In [199]:
df.describe()

год постройки  этажей в доме       position         price
count    8941.000000   14847.000000    14847.00000  1.484700e+04
mean     1990.177944      11.885566   545910.41995  6.156512e+06
std       135.043308       6.562196   497895.61106  4.725808e+06
min         0.000000       1.000000        5.00000  3.000000e+05
25%      1979.000000       5.000000       21.00000  3.800000e+06
50%      2010.000000      10.000000  1000000.00000  5.200000e+06
75%      2018.000000      17.000000  1000000.00000  7.000000e+06
max      2023.000000      40.000000  1000000.00000  1.900000e+08

Смотрим заполнение колонок

In [200]:
pd.set_option('display.max_rows', None)
df.notna().sum()

отделка                    1941
год постройки              8941
жилая площадь             11425
корпус, строение           1812
ремонт                    12653
тип комнат                 7659
общая площадь             14847
высота потолков            6908
пассажирский лифт          9943
количество комнат         14847
балкон или лоджия         11149
площадь кухни             13453
в доме                     4150
грузовой лифт              7842
официальный застройщик     2097
срок сдачи                 2097
вид сделки                 9583
тёплый пол                  952
техника                    3510
мебель                     4859
этажей в доме             14847
способ продажи            13620
парковка                   8411
тип дома                  14847
этаж                      14847
название новостройки       2097
санузел                   13188
двор                       8045
тип участия                2097
окна                      10593
запланирован снос            27
position

В датафрейме присутствуют колонки, плохо заполнены и восстановление пропусков не представляется возможным. Удалим их.

In [201]:
columns_for_del = [
    'отделка',
    'корпус, строение',
    'в доме','официальный застройщик',
    'тёплый пол','техника',
    'мебель',
    'название новостройки',
    'тип участия',
    'запланирован снос'
]
df.drop(columns_for_del,axis=1,inplace=True)

In [202]:
df.shape

(14847, 23)

In [203]:
df.rename(columns={'position': 'метро', 'price': 'цена'},inplace=True)

In [204]:
pd.set_option('display.max_rows', 10)
df

год постройки жилая площадь            ремонт      тип комнат  \
0               NaN           NaN   требует ремонта             NaN   
1               NaN           NaN     косметический   изолированные   
2               NaN           NaN               NaN             NaN   
3               NaN           NaN               NaN   изолированные   
4               NaN           NaN     косметический   изолированные   
...             ...           ...               ...             ...   
4919            NaN           NaN              евро   изолированные   
4920         2021.0       11.3 м²              евро             NaN   
4921         1970.0       43.1 м²     косметический   изолированные   
4922            NaN         12 м²     косметический             NaN   
4923            NaN         32 м²      дизайнерский   изолированные   

     общая площадь высота потолков пассажирский лифт количество комнат  \
0          37.4 м²             NaN               NaN                 1   
1          57.9 м²             NaN               NaN                 3   
2          36.8 м²             NaN               NaN                 1   
3            74 м²             NaN               NaN                 2   
4            48 м²             NaN               NaN                 2   
...            ...             ...               ...               ...   
4919       53.4 м²             NaN               NaN                 2   
4920       39.1 м²           2.6 м                 2                 1   
4921       58.7 м²             NaN               NaN                 4   
4922       33.4 м²             NaN               NaN            студия   
4923         66 м²           2.6 м               NaN                 2   

     балкон или лоджия площадь кухни грузовой лифт        срок сдачи  \
0               балкон       10.7 м²           NaN               NaN   
1               балкон          5 м²           NaN               NaN   
2                  NaN           NaN           NaN   2 кв. 2025 года   
3                  NaN       19.3 м²           NaN   1 кв. 2024 года   
4               балкон          9 м²           NaN               NaN   
...                ...           ...           ...               ...   
4919               NaN        8.3 м²           NaN               NaN   
4920               NaN       19.7 м²           нет               NaN   
4921            балкон        5.9 м²           NaN               NaN   
4922            балкон           NaN           NaN               NaN   
4923            лоджия         30 м²           NaN               NaN   

             вид сделки  этажей в доме способ продажи парковка     тип дома  \
0      возможна ипотека             10      свободная      NaN    кирпичный   
1      возможна ипотека              5      свободная      NaN    панельный   
2                   NaN             19            NaN      NaN   монолитный   
3                   NaN              5            NaN      NaN   монолитный   
4      возможна ипотека             10      свободная      NaN      блочный   
...                 ...            ...            ...      ...          ...   
4919                NaN              5      свободная      NaN    кирпичный   
4920   возможна ипотека             10      свободная      NaN    панельный   
4921   возможна ипотека              5      свободная      NaN    панельный   
4922                NaN              3      свободная      NaN    панельный   
4923   возможна ипотека             17      свободная      NaN    панельный   

           этаж       санузел двор                окна    метро      цена  
0       5 из 10           NaN  NaN                 NaN  1000000   4542000  
1        2 из 5           NaN  NaN                 NaN       11   4999000  
2       8 из 19           NaN  NaN                 NaN  1000000   4937472  
3        2 из 5           NaN  NaN                 NaN  1000000  10633526  
4       2 из 10           NaN  NaN                 NaN  1000000   5599

Смотрим все колонки, в названии которых присутствует слово "площадь"

In [205]:
column_list_square = list(filter(lambda item: "площадь" in item, df.columns.tolist()))
column_list_square

['жилая площадь', 'общая площадь', 'площадь кухни']

Смотрим корреляцию между площадями.

In [206]:
def clear_square(data):
    if data:
        return float(str(data).split()[0])
for col_name in column_list_square:
    df[col_name] = df[col_name].apply(clear_square)
df[column_list_square].corr()

жилая площадь  общая площадь  площадь кухни
жилая площадь       1.000000       0.891156       0.280022
общая площадь       0.891156       1.000000       0.498862
площадь кухни       0.280022       0.498862       1.000000

In [207]:
df[column_list_square].notna().sum()

жилая площадь    11425
общая площадь    14847
площадь кухни    13453
dtype: int64

Удаляем колонку "жилая площадь", т.к. она коррелирует с колонкой "общая площадь	".

In [208]:
df.drop(['жилая площадь'], axis=1,inplace=True)

Разбиваем колонку с категориальными признаками "метро" по one-hot кодированию на те, что небольше 15 мин до ближайшей станции и те, что больше.

In [209]:
from sklearn.preprocessing import OneHotEncoder

In [210]:
def metro_avaliable(value):
    if value<=15:
        return "шаговая доступность"
    else:
        return "не шаговая доступность"
df['метро'] = df['метро'].apply(metro_avaliable)
ohe = OneHotEncoder()
array_hot_encoded = ohe.fit_transform(df['метро'].values.reshape(-1, 1)).toarray()
df.drop(columns=['метро'], inplace=True)
data_hot_encoded = pd.DataFrame(array_hot_encoded, columns=[f"метро_{i}" for i in range(array_hot_encoded.shape[1])], index=df.index)
df = pd.concat([df, data_hot_encoded], axis=1)

Посмотрим какими данными заполнена колонка кол-во комнат.

In [211]:
pd.set_option('display.max_rows', None)
df['количество комнат'].value_counts()

 1                       5824
 2                       4909
 3                       2483
 студия                  1220
 4                        326
 5                         35
 свободная планировка      35
 6                          7
 7                          6
 8                          1
 9                          1
Name: количество комнат, dtype: int64

Удалим редкие данные, а "студия" заменим на 0. Количество комнат переводим в целочисленный признак.

In [212]:
for item in ['5','6','7','8','9','свободная планировка']:
    df.drop(df[df['количество комнат'].str.contains(item)].index, inplace = True)
def rep_count_room(value):
    if "студия" in value:
        return 0
    return int(value)
df['количество комнат']=df['количество комнат'].apply(rep_count_room)

Оценим корреляцию между признаками "количесвто комнат" и "общая площадь".

In [213]:
df[['общая площадь','количество комнат']].corr()

общая площадь  количество комнат
общая площадь           1.000000           0.758312
количество комнат       0.758312           1.000000

In [214]:
df['количество комнат'].value_counts()

1    5756
2    4846
3    2469
0    1204
4     322
Name: количество комнат, dtype: int64

Количество комнат коррелирует с общей площадью. К тому же количество данных разнится от колическтва комнат. Удалим колонку "количество комнат", оставив признак "общая площадь" как более описательный.

In [215]:
df.drop(['количество комнат'],axis=1,inplace=True)

Рассмотрим признаки "год постройки" и "срок сдачи".

In [216]:
pd.set_option('display.max_rows', 10)
df[['год постройки', 'срок сдачи']]

год постройки        срок сдачи
0               NaN               NaN
1               NaN               NaN
2               NaN   2 кв. 2025 года
3               NaN   1 кв. 2024 года
4               NaN               NaN
...             ...               ...
4919            NaN               NaN
4920         2021.0               NaN
4921         1970.0               NaN
4922            NaN               NaN
4923            NaN               NaN

[14597 rows x 2 columns]

В обоих колонках присутствуют пропуски. Оба признака можно объединить. Таким образом в колонке "год постройки" будут присутствовать будущие дома. Оставшиеся же пропуски данных пока что заполним текущим годом.

In [217]:
df['год постройки'].value_counts()

2021.0    691
2022.0    525
2017.0    429
2020.0    405
2016.0    380
         ... 
290.0       1
20.0        1
67.0        1
45.0        1
1920.0      1
Name: год постройки, Length: 116, dtype: int64

In [218]:
df['срок сдачи'].value_counts()

 сдан                                      497
 2 кв. 2024 года                           338
 1 кв. 2023 года                           205
 4 кв. 2023 года                           176
 4 кв. 2022 года                           168
                                          ... 
 2 кв. 2026 года                            14
 4 кв. 2025 года                            10
 4 кв. 2021 года, сдача не подтверждена      3
 2 кв. 2020 года, сдача не подтверждена      2
 2 кв. 2022 года, сдача не подтверждена      2
Name: срок сдачи, Length: 19, dtype: int64

In [219]:
# df = df[df['срок сдачи'].str.contains('сдача не подтверждена')==False]

In [220]:
# df.drop(df[df['срок сдачи'].str.contains('сдача не подтверждена')==True],axis=1,inplace=True)

Извдечем из строк признака "срок сдачи" год.

In [221]:
import re
import numpy as np
import datetime as dt

In [222]:
def extract_year(year, yearbuilding):
    if not np.isnan(year):
        year = int(year)
        if year>100:
            return year
        elif 22<year and year<100:
            return (1900+year)
        elif year<=22:
            return (2000+year)
    if type(yearbuilding)==str and yearbuilding.strip()=='сдан':
        return dt.datetime.now().year
    if type(yearbuilding)==str:
        return int(next(iter(re.findall(r"[0-9]{4,4}", yearbuilding))))
#     return -1

df['год постройки']=df.apply(lambda row: extract_year(row['год постройки'], row['срок сдачи']), axis=1)

In [223]:
df.shape

(14597, 22)

In [224]:
df['тип дома'].value_counts()

 кирпичный              5567
 монолитный             2800
 панельный              2767
 монолитно-кирпичный    2578
 блочный                 860
 деревянный               25
Name: тип дома, dtype: int64

In [225]:
df.drop(df[df['тип дома']==' деревянный'].index, inplace = True)

In [226]:
pd.set_option('display.max_rows', 10)
df['год постройки'].value_counts()

2022.0    1204
2021.0     700
2023.0     606
2024.0     572
2017.0     427
          ... 
1929.0       1
1898.0       1
1928.0       1
1860.0       1
290.0        1
Name: год постройки, Length: 108, dtype: int64

In [227]:
df['год постройки'].notna().sum()

10795

In [228]:
df = df[df['год постройки'].notna()]

Удалим года 217, 290.

In [229]:
df.drop(df[df['год постройки']==217].index, inplace = True)
df.drop(df[df['год постройки']==290].index, inplace = True)

In [230]:
df.shape

(10788, 22)

In [231]:
df['год постройки'].value_counts()

2022.0    1204
2021.0     700
2023.0     606
2024.0     572
2017.0     427
          ... 
1929.0       1
1898.0       1
1927.0       1
1928.0       1
1860.0       1
Name: год постройки, Length: 106, dtype: int64

In [232]:
# df = df.groupby('год постройки').filter(lambda x : len(x)>100)

Удаляем колонку "срок сдачи" за ненадобностью.

In [233]:
df.drop(['срок сдачи'],axis=1,inplace=True)

Ещё раз оцениваем заполненность данных.

In [234]:
pd.set_option('display.max_rows', None)
df.notna().sum()

год постройки        10788
ремонт                8703
тип комнат            5555
общая площадь        10788
высота потолков       5801
пассажирский лифт     7644
балкон или лоджия     8092
площадь кухни         9839
грузовой лифт         6160
вид сделки            7061
этажей в доме        10788
способ продажи        9583
парковка              6786
тип дома             10788
этаж                 10788
санузел               9695
двор                  6448
окна                  8368
цена                 10788
метро_0              10788
метро_1              10788
dtype: int64

Рассмотрим признак "этажей в доме" и этаж.

In [235]:
df['этажей в доме'].value_counts()

9     1597
5     1582
10    1087
17     901
16     901
19     636
3      533
18     516
24     420
25     415
6      282
14     226
4      196
7      194
8      188
20     187
12     185
2      157
22     144
23     111
21      83
15      81
11      72
13      31
26      30
1       20
27       9
37       2
38       1
34       1
Name: этажей в доме, dtype: int64

Удалим те данные, которые представлены мало.

In [236]:
# df = df.groupby('этажей в доме').filter(lambda x : len(x)>100)

In [237]:
# for item in [13, 1, 26, 27, 37, 34, 38]:
#     df.drop(df[df['этажей в доме']== item].index, inplace = True)

Рассмотрим колонку "этаж".

In [238]:
pd.set_option('display.max_rows', 10)
df['этаж'].value_counts()

 5 из 5      327
 3 из 5      322
 2 из 5      319
 4 из 5      314
 1 из 5      300
            ... 
 17 из 26      1
 14 из 21      1
 15 из 26      1
 13 из 26      1
 15 из 27      1
Name: этаж, Length: 349, dtype: int64

Из строк вычленим номер этажа.

In [239]:
df['этаж'] = df['этаж'].apply(lambda x: int(x.split()[0]))
df['этаж'].value_counts()

2     1390
3     1234
1     1187
5      973
4      972
      ... 
23      45
25      30
24      26
26       2
27       1
Name: этаж, Length: 27, dtype: int64

In [240]:
# df = df.groupby('этаж').filter(lambda x : len(x)>100)

Рассмотрим признак "обшая площадь".

In [241]:
pd.set_option('display.max_rows', 10)
df['общая площадь'].value_counts()

40.0     191
45.0     149
60.0     147
38.0     138
42.0     137
        ... 
89.3       1
102.2      1
82.4       1
82.7       1
104.8      1
Name: общая площадь, Length: 946, dtype: int64

In [242]:
df['общая площадь'] = df['общая площадь'].round()
df['общая площадь'].value_counts()

42.0     371
38.0     370
40.0     355
44.0     328
36.0     318
        ... 
131.0      1
171.0      1
128.0      1
155.0      1
153.0      1
Name: общая площадь, Length: 164, dtype: int64

Удалим редкие строки

In [243]:
# square_dict = dict(df['общая площадь'].value_counts())
# sk = square_dict.keys()
# for item in sk:
#     if square_dict[item]<17:
#         df.drop(df[df['общая площадь']==item].index, inplace = True)
# sub_df = df[df.groupby('city').city.transform('count')>3].copy() 
# df = df.groupby('общая площадь').filter(lambda x : len(x)>100)
# df['общая площадь'].value_counts()

Рассморим признак "ремонт".

In [244]:
df['ремонт'].value_counts()

 косметический      3227
 евро               2944
 требует ремонта    1609
 дизайнерский        923
Name: ремонт, dtype: int64

Заменим названия категорий и колонке числовыми по следующему принципу:
- требует ремонта = 1
- косметический = 2
- евро = 3
- дизайнерский = 4
- иначе = 5

In [245]:
def renovation(value):
    if value==" требует ремонта":
        return 1
    if value==" косметический":
        return 2
    if value==" евро":
        return 3
    if value==" дизайнерский":
        return 4
    return 5
df['ремонт']=df['ремонт'].apply(renovation)
df['ремонт'].value_counts()

2    3227
3    2944
5    2085
1    1609
4     923
Name: ремонт, dtype: int64

Рассмотрим колонку "высота потолков".

In [246]:
df['высота потолков'].value_counts()

 2.7 м     1849
 2.8 м     1024
 2.5 м      972
 2.6 м      878
 3 м        458
           ... 
 230 м        1
 18.4 м       1
 2750 м       1
 5 м          1
 25 м         1
Name: высота потолков, Length: 67, dtype: int64

Как видим, единицы измерения в некотрых местах нарушены. Исправим это, обработав колонку. Обработаем колонку "высота потолков". Для начала рассмотрим содержимое.

In [247]:
def extract_height(height):
    if type(height)==str:
        h = float(height.split()[0])
        if h<10:
            return h
        if h>10 and h<100:
            return h/10
        if h>100 and h<1000:
            return h/100
        if h>1000:
            return h/1000
        return h
    return height
df['высота потолков'] = df['высота потолков'].apply(extract_height)
df['высота потолков'].value_counts()

2.70    1889
2.80    1049
2.50     985
2.60     891
3.00     460
        ... 
2.77       1
2.63       1
4.19       1
3.28       1
1.46       1
Name: высота потолков, Length: 47, dtype: int64

Убираем "странные" значения

In [248]:
df.drop(df[df['высота потолков'].isin([1.46, 1.91, 1.20, 1.80, 1.84, 1.30, 0.30, 1.65, 1.50, 0.00])].index,inplace=True)

Дублирование данных.

In [249]:
df['год постройки'].value_counts()

2022.0    1191
2021.0     696
2023.0     602
2024.0     567
2017.0     425
          ... 
1929.0       1
1898.0       1
1927.0       1
1928.0       1
1860.0       1
Name: год постройки, Length: 106, dtype: int64

In [250]:
ldict = dict(df['год постройки'].value_counts())
lkeys = ldict.keys()
n = 1
for item in lkeys:
    level = ldict[item]
    if level>=400:
        continue
    if 300<=level and level<400:
        n = 2
    elif 200<=level and level<300:
        n = 3
    elif 100<=level and level<200:
        n = 4
    elif 80<=level and level<100:
        n = 5
    elif 60<=level and level<80:
        n = 7
    elif 40<=level and level<60:
        n = 10
    elif 30<=level and level<40:
        n = 15
    elif 20<=level and level<30:
        n = 20
    elif 10<=level and level<20:
        n = 25
    elif 8<=level and level<10:
        n = 50
    elif 5<=level and level<8:
        n = 90
    elif 2<=level and level<5:
        n = 200
    else:
        n = 400
    tmp = pd.concat([df[df['год постройки']==item]]*n)
    df = pd.concat([df,tmp])
    
df['год постройки'].value_counts()

2022.0    1191
2016.0    1140
2015.0    1041
2013.0     985
2025.0     975
          ... 
1955.0     390
2026.0     364
1952.0     364
1953.0     364
1951.0     260
Name: год постройки, Length: 106, dtype: int64

In [251]:
pd.set_option('display.max_rows', None)
df['высота потолков'].value_counts()

2.70    7567
2.50    6295
3.00    6268
2.60    5018
2.80    4340
2.90    1545
3.20    1518
3.10     739
3.50     735
4.00     428
3.60     427
4.50     205
4.30     201
2.10     182
3.40     149
2.40     148
3.70      85
3.30      82
2.00      49
2.20      42
2.75      42
2.30      34
4.20      31
2.65      30
2.55      29
6.00      22
2.85      15
4.80      14
2.77      11
4.19      11
3.28       8
2.34       5
1.84       5
5.00       5
2.64       4
2.72       3
2.63       3
2.74       2
3.90       2
6.60       1
Name: высота потолков, dtype: int64

Высота потолков.

In [252]:
ceiling_dict = dict(df['высота потолков'].value_counts())
ck = ceiling_dict.keys()
for item in ck:
    count = ceiling_dict[item]
    if count<1500:
        if 600<=count and count<=700:
            n = 2
        elif 400<=count and count<600:
            n = 3
        elif 300<=count and count<400:
            n = 4
        elif 200<=count and count<300:
            n = 6
        elif 100<=count and count<200:
            n = 12
        elif 90<=count and count<100:
            n = 15
        elif 80<=count and count<90:
            n = 17
        elif 70<=count and count<80:
            n = 20
        elif 60<=count and count<70:
            n = 25
        elif 50<=count and count<60:
            n = 30
        elif 40<=count and count<50:
            n = 35
        elif 30<=count and count<40:
            n = 45
        elif 20<=count and count<30:
            n = 55
        elif 10<=count and count<20:
            n = 120
        elif 8<=count and count<10:
            n = 150
        elif 5<=count and count<8:
            n = 250
        elif 2<=count and count<5:
            n = 600
        else:
            n = 1400
        tmp = pd.concat([df[df['высота потолков']==item]]*n)
        df = pd.concat([df,tmp])
#         df.drop(df[df['высота потолков']==item].index, inplace = True)
df['высота потолков'].value_counts()

3.10    1035339
3.50    1029735
2.70       7567
2.50       6295
3.00       6268
2.60       5018
2.80       4340
2.64       2404
2.10       2366
3.40       1937
2.40       1924
2.85       1815
2.63       1803
2.72       1803
2.00       1764
4.00       1712
3.60       1708
4.80       1694
2.55       1624
2.30       1564
2.90       1545
3.70       1530
3.20       1518
2.20       1512
2.75       1512
3.30       1476
4.50       1435
4.20       1426
4.30       1407
6.60       1401
2.65       1380
2.77       1331
4.19       1331
1.84       1255
2.34       1255
5.00       1255
6.00       1232
3.28       1208
3.90       1202
2.74       1202
Name: высота потолков, dtype: int64

Оценим заполненность признака.

In [253]:
print(f"непустые значения высоты потолков: {df['высота потолков'].notna().sum()}\nКол-во строк в датасете: {df.shape[0]}")

непустые значения высоты потолков: 2146093
Кол-во строк в датасете: 2169692


Заменим пустые значения в колонке "высота потолков" на среднее значение по всей колонке.

In [254]:
# df['высота потолков'].fillna(df['высота потолков'].mean(),inplace=True)
df = df[df['высота потолков'].notna()]
# df['высота потолков'].value_counts()

In [255]:
# df.drop(df[df['высота потолков']<0.0].index, inplace = True)
df['высота потолков'].notna().sum()

2146093

Смотрим колонку "тип дома"

In [256]:
df['тип дома'].value_counts()

 кирпичный              1936398
 монолитно-кирпичный     152805
 монолитный               35547
 панельный                17417
 блочный                   3926
Name: тип дома, dtype: int64

Удалим категорию "деревянный" за недостатком данных.

Заменим названия на числа.

In [257]:
def type_of_house(value):
    if value==" кирпичный":
        return 1
    if value==" монолитный":
        return 2
    if value==" панельный":
        return 3
    if value==" блочный":
        return 4
    return 5
df['тип дома']=df['тип дома'].apply(type_of_house)

Дублирование данных. В признаках "этаж", "этажей в доме", "общая площадь", "ремонт", "тип дома" продублируем те данные, котрые представлены мало.

Общая площадь

In [258]:
df['общая площадь'].value_counts()

35.0     287670
33.0     283351
26.0     164902
80.0     136254
76.0     133364
39.0     130461
79.0     127563
17.0     127526
38.0      89663
66.0      73206
122.0     71455
86.0      52536
49.0      38261
82.0      36587
44.0      32487
48.0      31408
40.0      30704
92.0      25502
182.0     15411
118.0     14033
45.0      13426
64.0      12982
56.0      11998
19.0      11444
125.0     11364
104.0     11268
42.0      10828
130.0      9811
57.0       9338
72.0       8837
73.0       8603
74.0       7330
24.0       7162
149.0      7005
34.0       6209
52.0       5699
230.0      5604
200.0      5604
21.0       4899
111.0      4576
85.0       4345
110.0      4224
152.0      4214
51.0       4190
25.0       3755
36.0       2322
37.0       2163
29.0       1951
32.0       1942
43.0       1813
62.0       1799
30.0       1735
15.0       1704
28.0       1642
100.0      1609
78.0       1488
105.0      1432
119.0      1415
190.0      1401
221.0      1401
128.0      1401
244.0      1401
54.0    

In [259]:
# ldict = dict(df['общая площадь'].value_counts())
# lkeys = ldict.keys()
# n = 1
# for item in lkeys:
#     count = ldict[item]
#     if count>100:
#         continue
#     if 50<=count and count<=100:
#         n = 2
#     elif 40<=count and count<50:
#         n = 3
#     elif 30<=count and count<40:
#         n = 4
#     elif 20<=count and count<30:
#         n = 5
#     elif 10<=count and count<20:
#         n = 10
#     elif 8<=count and count<10:
#         n = 15
#     elif 5<=count and count<8:
#         n = 20
#     elif 2<=count and count<5:
#         n = 50
#     else:
#         n=100
#     tmp = pd.concat([df[df['общая площадь']==item]]*n)
#     df = pd.concat([df,tmp])
    
# df['общая площадь'].value_counts()

Этажей в доме

In [260]:
df['этажей в доме'].value_counts()

4     678077
5     378808
3     291194
2     232401
1     219892
6     140229
10     48065
9      21748
24     20444
22     18411
18     14353
16     14219
19     12654
17     11094
23     10242
7       7435
11      7404
20      4971
34      4203
8       3049
14      2737
21      1785
13      1351
12       861
25       246
15       176
26        31
27         8
38         5
Name: этажей в доме, dtype: int64

In [261]:
ldict = dict(df['этажей в доме'].value_counts())
lkeys = ldict.keys()
n = 1
for item in lkeys:
    level = ldict[item]
    if level>=500:
        continue
    if 300<=level and level<=400:
        n = 2
    elif 200<=level and level<300:
        n = 3
    elif 100<=level and level<200:
        n = 4
    elif 80<=level and level<100:
        n = 5
    elif 60<=level and level<80:
        n = 7
    elif 40<=level and level<60:
        n = 10
    elif 30<=level and level<40:
        n = 15
    elif 20<=level and level<30:
        n = 20
    elif 10<=level and level<20:
        n = 25
    elif 8<=level and level<10:
        n = 50
    elif 5<=level and level<8:
        n = 90
    elif 2<=level and level<5:
        n = 200
    else:
        n = 400
    tmp = pd.concat([df[df['этажей в доме']==item]]*n)
    df = pd.concat([df,tmp])
    
df['этажей в доме'].value_counts()

4     678077
5     378808
3     291194
2     232401
1     219892
6     140229
10     48065
9      21748
24     20444
22     18411
18     14353
16     14219
19     12654
17     11094
23     10242
7       7435
11      7404
20      4971
34      4203
8       3049
14      2737
21      1785
13      1351
25       984
15       880
12       861
26       496
38       455
27       408
Name: этажей в доме, dtype: int64

Этаж

In [262]:
df['этаж'].value_counts()

2     860320
1     824562
5     151024
3     101252
4      64071
7      37806
17     17031
23     12663
18     10591
11     10413
14     10183
13      9507
9       8671
6       8191
22      5650
20      4313
8       3768
19      2785
21      1527
15      1494
10      1360
12       614
26       455
16       311
24       160
25       128
Name: этаж, dtype: int64

In [263]:
# ldict = dict(df['этаж'].value_counts())
# lkeys = ldict.keys()
# n = 1
# for item in lkeys:
#     level = ldict[item]
#     if level>400:
#         continue
#     if 300<=level and level<=400:
#         n = 2
#     elif 200<=level and level<300:
#         n = 3
#     elif 100<=level and level<200:
#         n = 4
#     elif 80<=level and level<100:
#         n = 5
#     elif 60<=level and level<80:
#         n = 7
#     elif 40<=level and level<60:
#         n = 10
#     elif 30<=level and level<40:
#         n = 15
#     elif 20<=level and level<30:
#         n = 20
#     elif 10<=level and level<20:
#         n = 25
#     elif 8<=level and level<10:
#         n = 50
#     elif 5<=level and level<8:
#         n = 90
#     elif 2<=level and level<5:
#         n = 200
#     else:
#         n = 400
#     tmp = pd.concat([df[df['этаж']==item]]*n)
#     df = pd.concat([df,tmp])
    
# df['этаж'].value_counts()

In [264]:
pd.set_option('display.max_rows', 10)
df['общая площадь'].value_counts()

35.0     287721
33.0     283355
26.0     164902
80.0     136260
76.0     133364
          ...  
143.0         5
124.0         5
153.0         5
113.0         4
464.0         1
Name: общая площадь, Length: 151, dtype: int64

In [265]:
df['высота потолков'].value_counts()

3.10    1035339
3.50    1029735
2.70       8496
3.00       6820
2.50       6367
         ...   
2.34       1255
6.00       1232
3.28       1208
3.90       1202
2.74       1202
Name: высота потолков, Length: 40, dtype: int64

In [266]:
df['этажей в доме'].value_counts()

4     678077
5     378808
3     291194
2     232401
1     219892
       ...  
15       880
12       861
26       496
38       455
27       408
Name: этажей в доме, Length: 29, dtype: int64

In [267]:
df['этаж'].value_counts()

2     860320
1     824562
5     151024
3     101252
4      64071
       ...  
12       614
26       455
16       311
24       160
25       128
Name: этаж, Length: 26, dtype: int64

Ремонт

In [268]:
df['ремонт'].value_counts()

2    947649
1    419258
4    417686
3    358059
5      6198
Name: ремонт, dtype: int64

In [269]:
# df = pd.concat([df,df[df['ремонт']==1]])
# df = pd.concat([df,df[df['ремонт']==5]])
# df = pd.concat([df,df[df['ремонт']==4]])
# df = pd.concat([df,df[df['ремонт']==4]])
# df['ремонт'].value_counts()

In [270]:
# ceiling_dict = dict(df['высота потолков'].value_counts())
# lkeys = ceiling_dict.keys()
# n = 1
# for item in lkeys:
#     count = ceiling_dict[item]
#     if count<400:
#         if count>50 and count<100:
#             n = 8
#         else:
#             n = 20
#         tmp = pd.concat([df[df['высота потолков']==item]]*n)
#         df = pd.concat([df,tmp])

Тип дома

In [271]:
df['тип дома'].value_counts()

1    1936635
5     153383
2      37291
3      17574
4       3967
Name: тип дома, dtype: int64

Удаляем колонку "вид сделки".

In [272]:
df.drop(['вид сделки'],axis=1,inplace=True)

In [273]:
df['цена за метр'] = df['цена']/df['общая площадь']

In [274]:
y = df['цена за метр']
features = [
            'год постройки',
            'общая площадь',
            'высота потолков',
            'ремонт',
            'тип дома',
            'этаж',
            'этажей в доме',
            'метро_0',
            'метро_1',
#             'лифт_0',
#             'лифт_1',
#             'окна во двор',
#             'окна на солнечную сторону',
#             'окна на улицу'
           ]
X = df[features]

In [275]:
from sklearn.linear_model import LinearRegression
from sklearn import tree
from sklearn.model_selection import train_test_split
from sklearn import metrics
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import cross_validate
from sklearn.ensemble import RandomForestRegressor
from sklearn.datasets import make_regression
from sklearn.ensemble import GradientBoostingRegressor

In [276]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=45)

Дерево решений

In [277]:
clf = tree.DecisionTreeRegressor()
scores = cross_validate(clf, X, y, cv=5, scoring=('r2', 'neg_mean_squared_error'), return_train_score=True)
scores

{'fit_time': array([1.29754114, 1.55582976, 1.53895569, 1.64457917, 1.67448115]),
 'score_time': array([0.04388046, 0.04786658, 0.04679918, 0.04886866, 0.05784392]),
 'test_r2': array([0.95007939, 1.        , 1.        , 1.        , 0.99999905]),
 'train_r2': array([0.99999979, 0.99999816, 0.99999813, 0.99999809, 0.99999835]),
 'test_neg_mean_squared_error': array([-1.80091999e+08, -3.41110641e-15, -5.70886959e-14, -5.65170024e-14,
        -3.86433023e+03]),
 'train_neg_mean_squared_error': array([ -966.08255697, -8183.94285268, -8183.94285268, -8183.94285268,
        -7217.86029571])}

In [278]:
print("%0.2f accuracy with a standard deviation of %0.2f" % (scores['test_r2'].mean(), scores['test_r2'].std()))

0.99 accuracy with a standard deviation of 0.02


In [279]:
clf = clf.fit(X_train, y_train)
y_pred = clf.predict(X_test)

ME = metrics.max_error(y_test, y_pred)
MAE= metrics.mean_absolute_error(y_test, y_pred)
MSE=metrics.mean_squared_error(y_test, y_pred)
RMSE= np.sqrt(MSE)
print(f"ME = {ME}\nMAE = {MAE}\nMSE = {MSE}\nRMSE = {RMSE}\nR2 score = {metrics.r2_score(y_test,y_pred)}")

ME = 176666.66666666666
MAE = 26.2760858801301
MSE = 1602200.9495243144
RMSE = 1265.7807667697887
R2 score = 0.9996368448756331


Случайный лес

In [280]:
regr = RandomForestRegressor(max_depth=2, random_state=0)
scores = cross_validate(clf, X, y, cv=5, scoring=('r2', 'neg_mean_squared_error'), return_train_score=True)
scores

{'fit_time': array([1.34341812, 1.64960408, 1.65157151, 1.8781836 , 1.72524118]),
 'score_time': array([0.04486918, 0.05087757, 0.05786061, 0.0578146 , 0.04886961]),
 'test_r2': array([0.9513594 , 1.        , 1.        , 1.        , 0.99999905]),
 'train_r2': array([0.99999979, 0.99999816, 0.99999813, 0.99999809, 0.99999835]),
 'test_neg_mean_squared_error': array([-1.75474275e+08, -3.41110641e-15, -5.70886959e-14, -5.65170024e-14,
        -3.86433023e+03]),
 'train_neg_mean_squared_error': array([ -966.08255697, -8183.94285268, -8183.94285268, -8183.94285268,
        -7217.86029571])}

In [281]:
print("%0.2f accuracy with a standard deviation of %0.2f" % (scores['test_r2'].mean(), scores['test_r2'].std()))

0.99 accuracy with a standard deviation of 0.02


In [282]:
regr = regr.fit(X_train, y_train)
y_pred = regr.predict(X_test)

ME = metrics.max_error(y_test, y_pred)
MAE= metrics.mean_absolute_error(y_test, y_pred)
MSE=metrics.mean_squared_error(y_test, y_pred)
RMSE= np.sqrt(MSE)
print(f"ME = {ME}\nMAE = {MAE}\nMSE = {MSE}\nRMSE = {RMSE}\nR2 score = {metrics.r2_score(y_test,y_pred)}")

ME = 383473.09059661784
MAE = 27455.74926163877
MSE = 2030188568.6557562
RMSE = 45057.613881071826
R2 score = 0.5398371331902766


Градиентный бустинг

In [283]:
params = {
    "n_estimators": 500,
    "max_depth": 4,
    "min_samples_split": 5,
    "learning_rate": 0.01,
    "loss": "squared_error",
}
reg = GradientBoostingRegressor(**params)
scores = cross_validate(reg, X, y, cv=5, scoring=('r2', 'neg_mean_squared_error'), return_train_score=True)
scores

{'fit_time': array([381.87964177, 433.77584839, 423.88286376, 378.34271979,
        405.54105711]),
 'score_time': array([2.07747245, 2.75071049, 1.97077537, 1.88099694, 1.90993571]),
 'test_r2': array([0.94799602, 0.97898248, 0.990242  , 0.98891538, 0.95906309]),
 'train_r2': array([0.98974163, 0.98307753, 0.98248642, 0.97897635, 0.98771796]),
 'test_neg_mean_squared_error': array([-1.87607930e+08, -7.78310731e+07, -4.44731035e+07, -4.86937253e+07,
        -1.66386413e+08]),
 'train_neg_mean_squared_error': array([-46199414.55309171, -75141202.39347307, -76804520.83947465,
        -89876081.23081839, -53846213.96947796])}

In [284]:
print("%0.2f accuracy with a standard deviation of %0.2f" % (scores['test_r2'].mean(), scores['test_r2'].std()))

0.97 accuracy with a standard deviation of 0.02


In [285]:
reg = reg.fit(X_train, y_train)
y_pred = reg.predict(X_test)

ME = metrics.max_error(y_test, y_pred)
MAE= metrics.mean_absolute_error(y_test, y_pred)
MSE=metrics.mean_squared_error(y_test, y_pred)
RMSE= np.sqrt(MSE)
print(f"ME = {ME}\nMAE = {MAE}\nMSE = {MSE}\nRMSE = {RMSE}\nR2 score = {metrics.r2_score(y_test,y_pred)}")

ME = 325050.83256572596
MAE = 4560.696810407571
MSE = 71426814.3376946
RMSE = 8451.438595747744
R2 score = 0.9838103867984627


<b><i>Квартира 1.</i></b><br>
<b>год постройки: </b>1972<br>
<b>общая площадь: </b>32.5<br>
<b>высота потолков: </b>2.7<br>
<b>ремонт: </b>косметический (2)<br>
<b>тип дома: </b>панельный (3)<br>
<b>этаж: </b>5<br>
<b>этажей в доме: </b>5<br>
<b>небольше 15 мин до метро: </b>0<br>
<b>больше 15 мин до метро: </b>1<br>
<br>
<font size=4><b>Цена за кв. метр: </b>116 923<br>
<b>Цена: </b>3 800 000<br><font>

In [287]:
xx = np.array([[1972,32.5,2.7,2,3,5,5,0,1]])
y_pre = reg.predict(xx)
print(f"Цена за кв. метр: {y_pre[0]}\nЦена: {y_pre[0]*32.5}")

Цена за кв. метр: 123035.90400211286
Цена: 3998666.880068668


C:\Users\Ilgiz\anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but GradientBoostingRegressor was fitted with feature names
  warnings.warn(


<b><i>Квартира 2.</i></b><br>
<b>год постройки: </b>2013<br>
<b>общая площадь: </b>51.3<br>
<b>высота потолков: </b>2.7<br>
<b>ремонт: </b>дизайнерский (4)<br>
<b>тип дома: </b>монолитно-кирпичный (5)<br>
<b>этаж: </b>7<br>
<b>этажей в доме: </b>17<br>
<b>небольше 15 мин до метро: </b>1<br>
<b>больше 15 мин до метро: </b>0<br>
<br>
<font size=4><b>Цена за кв. метр: </b>311 891<br>
<b>Цена: </b>16 000 000<br><font>

In [290]:
xx = np.array([[2013,51,2.7,4,5,7,17,1,0]])
y_pre = reg.predict(xx)
print(f"Цена за кв. метр: {y_pre[0]}\nЦена: {y_pre[0]*51}")

Цена за кв. метр: 166679.57977885785
Цена: 8500658.56872175


C:\Users\Ilgiz\anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but GradientBoostingRegressor was fitted with feature names
  warnings.warn(


<b><i>Квартира 3.</i></b><br>
<b>год постройки: </b>1969<br>
<b>общая площадь: </b>17.9<br>
<b>высота потолков: </b>2.5<br>
<b>ремонт: </b>косметический (2)<br>
<b>тип дома: </b>кирпичный (1)<br>
<b>этаж: </b>2<br>
<b>этажей в доме: </b>5<br>
<b>небольше 15 мин до метро: </b>0<br>
<b>больше 15 мин до метро: </b>1<br>
<br>
<font size=4><b>Цена за кв. метр: </b>162 011<br>
<b>Цена: </b>2 900 000<br><font>

In [291]:
xx = np.array([[1969,17.9,2.5,2,1,2,5,0,1]])
y_pre = reg.predict(xx)
print(f"Цена за кв. метр: {y_pre[0]}\nЦена: {y_pre[0]*17.9}")

Цена за кв. метр: 127874.82046885866
Цена: 2288959.2863925695


C:\Users\Ilgiz\anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but GradientBoostingRegressor was fitted with feature names
  warnings.warn(


In [292]:
xx = np.array([[2013,60,2.7,3,1,8,9,0,1]])
y_pre = reg.predict(xx)
print(f"Цена за кв. метр: {y_pre[0]}\nЦена: {y_pre[0]*60}")

Цена за кв. метр: 124178.73379824356
Цена: 7450724.027894613


C:\Users\Ilgiz\anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but GradientBoostingRegressor was fitted with feature names
  warnings.warn(
